src/sandstorm/sandstorm-ip-bridge.c++

```diff
@@ -0,0 +1,181 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2014 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program is useful for including in Sandstorm application packages where
+// the application itself is a legacy HTTP web server that does not understand
+// how to speak the Cap'n Proto interface directly.  This program will start up
+// that server and then redirect incoming requests to it over standard HTTP on
+// the loopback network interface.
+
+#include "sandstorm/sandstorm-ip-bridge.h"
+
+#include <kj/debug.h>
+#include <netinet/ip.h>
+#include <linux/netfilter_ipv4.h>
+#include <unistd.h>
+
+namespace sandstorm {
+  namespace ipbridge {
+
+  class RefcountedAsyncIoStream: public kj::Refcounted {
+  public:
+    explicit RefcountedAsyncIoStream(kj::Own<kj::AsyncIoStream>&& stream)
+        : stream(kj::mv(stream)) {}
+    kj::Own<kj::AsyncIoStream> stream;
+  };
+
+  static const int BUFFER_SIZE = 8192;
+
+  // TCP handling
+  struct AcceptedConnection {
+    class Downstream final: public ByteStream::Server {
+    private:
+      kj::Own<RefcountedAsyncIoStream> connection;
+
+    public:
+      explicit Downstream(kj::Own<RefcountedAsyncIoStream> && connection) :
+        connection(kj::mv(connection)) {}
+
+      kj::Promise<void> write(WriteContext context) {
```

Comment:
The ByteString interface is documented to allow multiple write()s at a time (to avoid starving the stream between one write() returning and the next starting), but currently AsyncIoStream doesn't like overlapping writes. I would like to fix `AsyncIoStream` to allow this (mostly because it fits better with Win32's system calls, though there's not much benefit on Linux), but haven't gotten around to that yet.

For now, `Downstream` should have a `Promise<void>` member which is the promise returned by the last call to write(), and before issuing a new write() you should wait for the previous one to complete. You'll also need to `fork()` these promises so that you have something to return to the direct caller.

In `done()`, you'll want to make sure not to call `shutdownWrite()` until the last `write()` has completed.